### DISCUSSION TITLE
Why `Span` has an `enter()` method?

### DISCUSSION THREAD

**Discussion by psandana** [tokio-rs/tracing#2989] [Q&A] [answered]
Why does this `enter()` method exists? 
https://github.com/tokio-rs/tracing/blob/master/tracing/src/span.rs#L783

Can it be the `Span` object itself can hold the lifetime of the span?

**Comment by mladedav** [comment, marked as answer]
Because sometimes you want to enter it sometime later after its creation and sometimes you want to enter and exit it multiple times (for example instrumented futures do this).

**Reply by psandana** [reply]
Interesting point the latest. For the first I would expect a scope would be enough. 

Thanks, I will explore more about that pattern.

**Reply by mladedav** [reply]
The problem with scopes is you need to have all fields available where you want to create the span.

For example

```rust
let span = info_span!("name", ?local_variable);
do_stuff.instrument(span).await;
```

becomes impossible and if you want to have the local variable in your span, you would have to pass it along.
